### ISSUE TITLE
Should project references check for transitive references to do the declaration file include check

### ISSUE THREAD

**Issue by sheetalkamat** [Microsoft/TypeScript#27200]
Issue created from https://github.com/Microsoft/TypeScript/pull/27195#pullrequestreview-156553913
Need to identify what needs to happen in this scenario:

Does this work for transitive projectReferenes? I'm asking because I struggled with that too.

```ts
// @filename: tsconfig.a.json
{"compilerOptions": {"composite"}, "files": ["a.ts"]}

// @filename: a.ts
export class A {}

// @filename: tsconfig.b.json
{"compilerOptions": {"composite"}, "files": ["b.ts"], "references": [{"path": "tsconfig.a.json"}]}

// @filename: b.ts
import {A} from './a';
export const b = new A();

// @filename: tsconfig.c.json
{"files": ["c.ts"], "references": [{"path": "tsconfig.b.json"}]}

// @filename: c.ts
import {b} from './b';
console.log(b);
```

Compiling `tsconfig.c.json` should resolve imports as `b.d.ts` and `a.d.ts`. I guess the current implementation resolves imports to `b.d.ts` and `a.ts`.

**Comment by ajafff** [comment]
I verified that this is a problem in the newest nightly release. After all these recent refactorings, it resovles to the original files of transitive project references, even if `outDir` option is used. Previously this was only an issue if `ourDir === rootDir`.

**Comment by ajafff** [comment]
@RyanCavanaugh @sheetalkamat I assume the changes that lead to this issue landed/will land in ts@3.1. Will there be a fix for this issue in ts@3.1? The current behavior regarding transitive project references defeats the whole purpose of the feature. This is especially a problem in big projects with a deep hierarchy of project references.

* compiler needs to parse and check the source of transitive project references
* different `compilerOptions` in referenced projects may cause unexpected compiler errors
* `--watch` will rebuild for changes that don't affect emitted declaration files
